spec/jobs/validate_moab_job_spec.rb

```diff
@@ -0,0 +1,143 @@
+                                                                    note: a_string_starting_with(exp_str))
+    end
+
+    it 'reports failure to workflow server when there are validation errors' do
+      job.perform(druid)
+      validation_err_substring = 'druid:bj102hs9687-v0001: version_additions: file_differences'
+      expected_str_regex = /^Problem with Moab validation run on .*#{validation_err_substring}.*/
+      expect(workflow_client).to have_received(:update_error_status).with(druid: druid,
+                                                                          workflow: 'preservationIngestWF',
+                                                                          process: 'validate-moab',
+                                                                          error_msg: a_string_matching(expected_str_regex))
+    end
+
+    context 'when validation runs' do
+      let(:storage_object_validator) { Stanford::StorageObjectValidator.new(moab) }
+      let(:storage_object_version1) { Moab::StorageObjectVersion.new(moab, 1) }
+      let(:storage_object_version2) { Moab::StorageObjectVersion.new(moab, 2) }
+      let(:storage_object_version3) { Moab::StorageObjectVersion.new(moab, 3) }
+      let(:verification_result) { Moab::VerificationResult.new('ignore_here') }
+
+      before do
+        allow(Stanford::StorageObjectValidator).to receive(:new).and_return(storage_object_validator)
+        allow(storage_object_validator).to receive(:validation_errors).and_call_original
+        allow(verification_result).to receive(:verified).and_return(true)
+        allow(Moab::VerificationResult).to receive(:new).and_return(verification_result)
+        allow(moab).to receive(:version_list).and_return([storage_object_version1, storage_object_version2, storage_object_version3])
+        allow(storage_object_version1).to receive(:verify_signature_catalog).and_call_original
+        allow(storage_object_version1).to receive(:verify_manifest_inventory).and_call_original
+        allow(storage_object_version1).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version1).to receive(:verify_version_inventory).and_call_original
+        allow(storage_object_version2).to receive(:verify_signature_catalog).and_call_original
+        allow(storage_object_version2).to receive(:verify_manifest_inventory).and_call_original
+        allow(storage_object_version2).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version2).to receive(:verify_version_inventory).and_call_original
+        allow(storage_object_version3).to receive(:verify_signature_catalog).and_call_original
+        allow(storage_object_version3).to receive(:verify_manifest_inventory).and_call_original
+        allow(storage_object_version3).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version3).to receive(:verify_version_inventory).and_call_original
```

Comment:
i'm _slightly_ uncomfortable with how closely tied this section of tests is to the current implementation (in terms of the new class' calls to moab-versioning methods and in terms of assumptions about calls made by those moab-versioning methods).

on the other hand, we know those methods do what we want, they're not super likely to change soon, and the app code does call them out in the comments.

i think i was hoping that the tests would make more use of the existing fixture objects with known problems, but looking back at `checksum_validator_spec.rb`, it does more of this sort of mocking and reaching into moab-versioning than i remembered, and it's a little less clear about which druids have which issues than i'd hoped/remembered.

still... how would you feel about adding a test or two that just runs on a problematic moab, mocking none of the moab-versioning calls, with some expectations about what specific error messages get sent to WF service?  maybe even just a more exacting message expectation in the test on line 47?  and maybe one good moab resulting in marking the WF step complete?

for example, i think `zz111rr1111` has a checksum mismatch for `v0001/data/content/eric-smith-dissertation-augmented.pdf`, `tt222tt2222` is missing a content file, `zz333vv3333` is missing a sigCat, and `bz514sm9647` is ok?  that's my read from looking at the other checksum validator spec for a bit.

regardless, i don't think i'd block merging over this point.  i believe this code works and pushes the ball forward, just worried about such tightly coupled tests making it harder to refactor later.

Reply:
In a perfect world, I agree with you.  I took the expeditious route.  As you note, these tests aren't about showing that moab-validation can trap errors, they're about showing that we actually perform checksum validations, using the moab-val methods.

I did think about your point when I was writing the tests, and felt that since I was testing that the moab-val methods which do the actual comparison work were called (and not the moab-val method that is called in the job) ... that this was a reasonable middle ground.

Checking for specific errors ... could end up depending on the strings used by moab-val for error messages, or on specific errors thrown.   At some point, you have to trust that the tools you're using ... work.   If moab-val changes significantly, I think we'll know that prescat will need work.

Reply:
indeed